### ISSUE TITLE
`unsound-return-statement` flags regex match group return

### ISSUE THREAD

**Issue by sharkdp** [astral-sh/ty#4578]
ty's strict `unsound-return-statement` rule flags this code:

```py
import re

def filename(text: str) -> str:
    match = re.search(r'"(a+)"', text)
    if match:
        return match.group(1)  # Unsound return statement: `str | Any` is not a subtype of `str`
    else:
        return ""
```

```
Unsound return statement: `str | Any` is not a subtype of `str` (unsound-return-statement) [Ln 6, Col 16]
Expected a subtype of `str` because of the return type [Ln 3, Col 28]
info: `str | Any` is assignable to `str`, but not a subtype of `str`
info: element `Any` of union `str | Any` is not a subtype of `str`
help: Consider using an `assert` to narrow the type prior to the `return` statement
```

This is due to typeshed's [unfortunate annotation on `re.Match.group`](https://github.com/python/typeshed/issues/9465).

I am opening a ty discussion anyway because it seems like other type checkers allow this code to pass, even with their corresponding strict checks enabled. It seems like they flag a return if the type is `Any`, but not if it's `str | Any`.

Reading the `info` messages in ty's output (subtyping vs assignability) tell me that this is probably intentional, but I wanted to make sure that that's the case. FYI @AlexWaygood

**Comment by AlexWaygood** [comment]
Yes, this is intentional.

Typeshed's annotation here is `AnyStr | Any` rather than `AnyStr | None` because there are many regular expressions in real code, such as this one, where you know (because of the nature of the regex) that `Match.group()` will _never_ return `None` and so it would be very "annoying" for these users to have to explicitly account for the possibility that `Match.group()` returns `None`. But there are certainly different regexes to the above example where `Match.group()` can return `None` -- any regex with an optional group:

```pycon
>>> import re
...
... def filename(text: str) -> str:
...     match = re.search(r'(a+)?', text)
...     if match:
...         return match.group(1)
...     else:
...         return ""
...
>>> filename("") is None
True
```

The premise of this opt-in, disabled-by-default check is that users who want their type checker to actually ensure soundness will want to know if they have a fully static return annotation but a value returned from that function inhabits a type that is not actually a subtype of that fully static return annotation. This case seems like an _excellent_ example of why you might want that: ty cannot statically verify whether `match.group()` will actually return a subtype of `str` unless you add an `assert` statement like this, which is enough to make the error go away:

```py
import re

def filename(text: str) -> str:
    match = re.search(r'"(a+)"', text)
    if match:
        group = match.group(1)
        assert isinstance(group, str)
        return group
    else:
        return ""
```

If you don't add that `assert` statement, and you have a regex pattern that has an optional group in it, there's a very real chance that `None` could escape from your fully-static function that's declared as returning `str`.

> I am opening a ty discussion anyway because it seems like other type checkers allow this code to pass, even with their corresponding strict checks enabled

Right, and IMO it's kind of cool that we catch this unsoundness when mypy's equivalent check does not! I designed the check like this deliberately, to avoid having the same shortcoming.

**Comment by AlexWaygood** [comment]
We could, of course, add extensive special casing for regex `Match` patterns specifically, so that if the pattern originates from a literal string we're able to infer precisely whether a `.group()` call can be guaranteed never to return `None`. But the mypy PR for that idea has been open for 7 years now, suggesting that this might not be trivial to do. https://github.com/python/mypy/pull/7803. And I'm not sure that this should be high priority, given that the `unsound-return-statement` diagnostic can be easily worked around here by simply adding an `assert`.

**Comment by sharkdp** [comment]
Agree with everything you said! Thank you.

I understand why typeshed annotates it this way. I said "unfortunate" because if those stdlib APIs were to be redesigned with Python typing in mind, then there would certainly be ways to make more users happy. For example, there could be a "I know what I'm doing" function that would do the not-`None` runtime validation/assertion for you, which then could return `str` / `AnyStr`.
